### ISSUE TITLE
Prevent returning password field from SCIM

### ISSUE THREAD

**Issue by kuldeepaggarwal** [RIPAGlobal/scimitar#56]
`Scimitar::Schema::User` has a password attribute:

```rb
Attribute.new(name: 'password', type: 'string', mutability: 'writeOnly', returned: 'never'),
```

and as per the definition, it should never be returned in the response. However, if client sends a password and Service provider stores the password using the `scim_attributes_map` then SCIMitar is responding back with the password value.

I believe, the issue is in `Scimitar::Resources::Base#as_json`, when we are calling 
https://github.com/RIPAGlobal/scimitar/blob/914c0e10d21b66eaee4e85b5f4cd416528b0ee4c/app/models/scimitar/resources/base.rb#L143

without excluding those files which has a property `returned='never'`.

### Proposed Solution

```rb
def non_returnable_attributes
  self.class.schemas.flat_map(&:scim_attributes).select { |attribute| attribute.returned == 'never' }.map(&:name)
end

def as_json(options = {})
  ...
  original_hash = super(options).except('errors', *non_returnable_attributes)
  ...
end
```

**Comment by kuldeepaggarwal** [comment]
Sorry to hear that you are affected with flu, @pond! Thanks for acknowledging this as a bug. I just realized this hack won't work unless one more change. Here is the monkey patch that I added in my application when I completed E2E testing.

https://github.com/RIPAGlobal/scimitar/blob/9d183fd42bf3d07278054e3f670de1110a78dcab/app/controllers/scimitar/active_record_backed_resources_controller.rb#L80

```rb
module ScimitarResourceJSONOverride
  def as_json(options = {})
    return super unless options.key?(:layout) # this change is needed to parse password SCIM params from request when passing from Controller to UserResource else password will not be passed from Controller to model.

    non_returnable_attributes = self.class.schemas.flat_map(&:scim_attributes).filter_map { |attribute| attribute.name if attribute.returned == 'never' }
    super.except(*non_returnable_attributes)
  end
end


Scimitar::Resources::Base.include(ScimitarResourceJSONOverride)
```

**Comment by pond** [comment]
@kuldeepaggarwal This took far too long, sorry. Once I got back there was a backlog of work and things snowballed from there.

https://github.com/RIPAGlobal/scimitar/pull/80 implements (I think!) your initial suggestion, but without the amendment you have patched in with https://github.com/RIPAGlobal/scimitar/issues/56#issuecomment-1590153772 - I don't think it's necessary (but don't entirely understand it so could be very wrong). Please can you check the PR and see if that looks OK, and/or otherwise elaborate on the `:layout` key check you made in the monkey patch?

**Comment by pond** [comment]
@kuldeepaggarwal This should now be fixed via 2.6.0/2.6.1, or 1.7.0/1.7.1 all now on RubyGems. Please let me know if you agree and, if so, we can close this issue `:-)`

**Comment by pond** [comment]
I think it's been long enough on this to just go ahead and close it. If you see problems, please open a new issue. Thanks!
